From 88f36d2a574a9e540f87056cc9d0726967d5d1bd Mon Sep 17 00:00:00 2001
From: Fancy Fang <chen.fang@nxp.com>
Date: Mon, 7 Aug 2017 12:18:49 +0800
Subject: [PATCH 2336/5242] MLK-16136-15 video: fbdev: dcss: enable pan
 display function

commit  bf9376b8a99ff21675d8613c359654b92a2d9f90 from
https://source.codeaurora.org/external/imx/linux-imx.git

Enable pan display function for dcss framebuffer driver.

Signed-off-by: Fancy Fang <chen.fang@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/video/fbdev/mxc/imx_dcss.c |   44 +++++++++++++++++++++++++++++++++++-
 1 file changed, 43 insertions(+), 1 deletion(-)

diff --git a/drivers/video/fbdev/mxc/imx_dcss.c b/drivers/video/fbdev/mxc/imx_dcss.c
index 0e794b0..21ef6435 100644
--- a/drivers/video/fbdev/mxc/imx_dcss.c
+++ b/drivers/video/fbdev/mxc/imx_dcss.c
@@ -2826,8 +2826,50 @@ static int dcss_blank(int blank, struct fb_info *fbi)
 }
 
 static int dcss_pan_display(struct fb_var_screeninfo *var,
-		struct fb_info *fbi)
+			    struct fb_info *fbi)
 {
+	int ret = 0;
+	uint32_t offset, luma_addr, chroma_addr = 0;
+	struct dcss_channel_info *cinfo = fbi->par;
+	struct dcss_info *info = cinfo->dev_data;
+	struct platform_device *pdev = info->pdev;
+	struct cbuffer *cb = &cinfo->cb;
+	struct dcss_pixmap *input = &cinfo->input;
+
+	/* TODO: change framebuffer memory start address */
+	luma_addr = var->reserved[0] ? var->reserved[0] :
+				       fbi->fix.smem_start;
+
+	/* change display offset in framebuffer */
+	if (var->xoffset > 0) {
+		dev_dbg(&pdev->dev, "x panning not supported\n");
+		return -EINVAL;
+	}
+
+	if ((var->yoffset + var->yres > var->yres_virtual)) {
+		dev_err(&pdev->dev, "y panning exceeds\n");
+		return -EINVAL;
+	}
+
+	offset = fbi->fix.line_length * var->yoffset;
+
+	fill_sb(cb, cinfo->dpr_addr + 0xc0, luma_addr + offset);
+
+	/* Two planes YUV format */
+	if (fmt_is_yuv(input->format) == 2) {
+		chroma_addr = luma_addr +
+			      var->xres * var->yres * (var->bits_per_pixel >> 3);
+		fill_sb(cb,
+			cinfo->dpr_addr + 0x110,
+			chroma_addr + (offset >> 1));
+	}
+
+	ret = commit_to_fifo(fbi->node, info);
+	if (ret) {
+		dev_err(&pdev->dev, "commit config failed\n");
+		return ret;
+	}
+
 	return 0;
 }
 
-- 
1.7.9.5

